src/drivers/intel/cavs/ssp.c

```diff
@@ -345,102 +346,57 @@ static inline int ssp_set_config(struct dai *dai,
-		ret = -EINVAL;
-		goto out;
-	}
-#elif defined(CONFIG_CANNONLAKE)
-	if (!config->ssp.mclk_rate || config->ssp.mclk_rate > F_24000_kHz) {
-		trace_ssp_error("ssp_set_config() error: "
-				"invalid MCLK = %d Hz (valid < 24000kHz)",
-				config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-	if (!config->ssp.bclk_rate ||
-	    config->ssp.bclk_rate > config->ssp.mclk_rate) {
-		trace_ssp_error("ssp_set_config() error: "
-				"BCLK %d Hz = 0 or > MCLK %d Hz",
-				config->ssp.bclk_rate, config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
+		if (ssp_freq[i].freq % config->ssp.mclk_rate == 0) {
+			clk_index = i;
+			clk_set = 1;
+		}
 	}
 
-	if (F_24000_kHz % config->ssp.mclk_rate == 0) {
-		mdivr_val = F_24000_kHz / config->ssp.mclk_rate;
+	if (clk_set) {
+		mdivc |= MCDSS(ssp_freq[clk_index].enc);
+		mdivr_val = ssp_freq[clk_index].freq / config->ssp.mclk_rate;
 	} else {
-		trace_ssp_error("ssp_set_config() error: 24.0MHz / %d Hz MCLK not divisable",
+		trace_ssp_error("ssp_set_config() error: MCLK %d",
 				config->ssp.mclk_rate);
 		ret = -EINVAL;
 		goto out;
 	}
 
-	if (F_24000_kHz % config->ssp.bclk_rate == 0) {
-		mdiv = F_24000_kHz / config->ssp.bclk_rate;
-	} else {
-		trace_ssp_error("ssp_set_config() error: 24.0MHz / %d Hz BCLK not divisable",
-				config->ssp.bclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-#else
-	if (!config->ssp.mclk_rate || config->ssp.mclk_rate > F_24576_kHz) {
-		trace_ssp_error("ssp_set_config() error: "
-				"invalid MCLK = %d Hz (valid < 24567kHz)",
-				config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-	if (!config->ssp.bclk_rate ||
-	    config->ssp.bclk_rate > config->ssp.mclk_rate) {
-		trace_ssp_error("ssp_set_config() error: "
-				"BCLK %d Hz = 0 or > MCLK %d Hz",
-				config->ssp.bclk_rate, config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
-	}
-	if (F_24576_kHz % config->ssp.mclk_rate == 0) {
-		/* select Audio Cardinal clock for MCLK */
-		mdivc |= MCDSS(1);
-		mdivr_val = F_24576_kHz / config->ssp.mclk_rate;
-	} else if (config->ssp.mclk_rate <= F_19200_kHz &&
-		   F_19200_kHz % config->ssp.mclk_rate == 0) {
-		mdivr_val = F_19200_kHz / config->ssp.mclk_rate;
-	} else {
-		trace_ssp_error("ssp_set_config() error: MCLK %d",
-				config->ssp.mclk_rate);
-		ret = -EINVAL;
-		goto out;
+	/* BCLK config */
+	/* searching the smallest possible bclk source */
+	clk_set = 0;
+	for (i = ARRAY_SIZE(ssp_freq) - 1; i >= 0; i--) {
+		if (config->ssp.bclk_rate > ssp_freq[i].freq)
+			break;
+
+		if (ssp_freq[i].freq % config->ssp.bclk_rate == 0) {
+			clk_index = i;
+			clk_set = 1;
+		}
 	}
 
-	if (F_24576_kHz % config->ssp.bclk_rate == 0) {
-		/* select Audio Cardinal clock for M/N dividers */
-		mdivc |= MNDSS(1);
-		mdiv = F_24576_kHz / config->ssp.bclk_rate;
-		/* select M/N output for bclk */
-		sscr0 |= SSCR0_ECS;
-	} else if (F_19200_kHz % config->ssp.bclk_rate == 0) {
-		mdiv = F_19200_kHz / config->ssp.bclk_rate;
+	if (clk_set) {
+		mdivc |= MNDSS(ssp_freq[clk_index].enc);
```

Comment:
this is odd, the encoding is 0,1,2 so wondering if the macro definition works here:
include/sof/ssp.h:#define MNDSS(x)      ((x) << 20)

Reply:
As I found in specs we can set MNDSS (bits 21:20) and MCDDS (bits: 17:16) with:
```
00: XTAL Oscillator clock
01: Audio Cardinal clock
10: Audio PLL fixed clock
11: Reserved
```
Defines in sof/clk.h covers that values:
```
#define CLOCK_SSP_XTAL_OSCILLATOR		0x0
#define CLOCK_SSP_AUDIO_CARDINAL		0x1
#define CLOCK_SSP_PLL_FIXED			0x2
```
As I checked that macros should work correctly.